Draft projects musn't be publicly accessible

lutangar opened the pull request:
> Fix #220 

> Most of this PR is quite off topic but since e2e tests weren't launched in a while, I had to carry out some refactoring :
> - use existing environment variables in cypress instead of hard coded credentials
> - leverage __Cypress__ session mechanism to cache authentication requests
> - fix some tests which could failed when run in parallel (by enforcing tests isolations).
>
> Hopefully it won't happens again since e2e tests are now run in CI.

Forbid access to draft projects from any users except owner & admin.

lutangar committed:
chore(e2e): fix e2e tests

lutangar committed:
fix(user): a user may reuse its own email

lutangar committed:
chore(user): force active property to be a boolean

lutangar committed:
chore(cypress): fix cypress config

lutangar committed:
chore(e2e): run e2e tests on CI

lutangar committed:
chore(auth): remove useless tests

lutangar committed:
fix(project): forbid requesting a draft project

fix(project): forbid requesting a draft project

fix(project): fix project guard

lutangar committed:
fix(project): a visitor may see all non-draft projects

...fetching projects was returning a 500...

bastiendmt commented on apps/backend-e2e/src/e2e/auth.cy.ts:
```diff
@@ -15,140 +15,87 @@ describe('AUTHENTIFICATION TESTS', () => {
-      TESTS
-   */
-  it('Auth -> Connecting with incorrect user.', () => {
+  it('Should fail when trying to connect without any credentials 1/2.', () => {
```
having the same test name with `X/2` is not very descriptive, I think you should instead specify the differences between both tests. Here calling GET vs POST

lutangar replied:
I agree ! But this serie of test was just moved from `user.cy.ts` to `auth.cy.ts` so I won't refactor this here.
I won't dig on these specific tests since I don't find them very useful, but overall I'd prefer them to describe testing scenario rather than something too close to implementation details.

bastiendmt commented on apps/backend-e2e/src/e2e/auth.cy.ts:
```diff
@@ -15,140 +15,87 @@ describe('AUTHENTIFICATION TESTS', () => {
     });
   });
-  it('Auth -> Check self-profile of admin user with correct jwt.', () => {
+  it('Connecting with incorrect user.', () => {
```
```suggestion
  it('Should fail to connect with incorrect user.', () => {
```

Sentence more descriptive : 
- follows test naming convention `it should ...`
- explicitly indicates the result

lutangar replied:
As I said earlier this code was juste moved over (and disclaimer -- I'm not the author :sweat_smile: )
But you're right about the naming convention : it's the idiomatic way of chai, etc.
However If I had to refactor this, I would instead describe testing scenario rather than being too close to implementation details and __even if__ it breaks the `it` convention. 
This convention is inherited from __unit testing__ library and is too restrictive to describe something closer to a user story.

bastiendmt commented on apps/backend-e2e/src/e2e/auth.cy.ts:
```diff
@@ -15,140 +15,87 @@ describe('AUTHENTIFICATION TESTS', () => {
     });
   });
-  it('Auth -> Connecting with proper editor user but wrong password.', () => {
+  it('Connecting with proper user admin but wrong admin password.', () => {
```
```suggestion
  it('Should fail to connect as an admin with incorrect password.', () => {
```

bastiendmt commented on apps/backend-e2e/src/e2e/auth.cy.ts:
```diff
@@ -15,140 +15,87 @@ describe('AUTHENTIFICATION TESTS', () => {
     });
   });
-  it('Auth -> Connecting correctly with editor user.', () => {
+  it('Login as admin', () => {
```
```suggestion
  it('Should connect as admin', () => {
```

Be consistent with terms used `Login/Connecting`

lutangar replied:
If I had to refactor this, I would rather use "login" instead of "connect".
This could be the subject of a new issue tho.

lutangar replied:
https://github.com/datatlas-erasme/datatlas/issues/263

bastiendmt commented on apps/backend-e2e/src/e2e/auth.cy.ts:
```diff
@@ -15,140 +15,87 @@ describe('AUTHENTIFICATION TESTS', () => {
+  it('Login as admin', () => {
+    cy.login(Cypress.env('admin_credentials'));
+  });
+  it('Connecting with proper editor user but wrong password.', () => {
```
```suggestion
  it('Should fail to connect as an editor with incorrect password.', () => {
```

bastiendmt commented on apps/backend-e2e/src/e2e/auth.cy.ts:
```diff
@@ -15,140 +15,87 @@ describe('AUTHENTIFICATION TESTS', () => {
-      expect(response.body.active).to.eq(true);
-      expect(response.body.email).to.eq('editor');
-    });
+  it('Login as an editor', () => {
```
```suggestion
  it('Should connect as an editor', () => {
```

lutangar replied:
https://github.com/datatlas-erasme/datatlas/issues/263

bastiendmt commented on apps/backend-e2e/src/e2e/auth.cy.ts:
```diff
@@ -15,140 +15,87 @@ describe('AUTHENTIFICATION TESTS', () => {
-      expect(response.body.email).to.eq('editor');
-    });
+  it('Login as an editor', () => {
+    cy.login(Cypress.env('editor_credentials'));
```
No expect here ?

lutangar replied:
This is a Cypress command, if the command fails, so does the test, so an expect statement would be superfluous.

bastiendmt commented on apps/backend-e2e/src/e2e/project.cy.ts:
```diff
@@ -526,72 +490,69 @@ describe('PROJECT ACTIONS', () => {
   });
 
-  // UODATE
+  // UPDATE
```
Have you considered using describe block instead of comments for separating tests ?

lutangar replied:
Sure I'd love more describe block instead of comments !

> Just so you know, some tests are still dependent on each others -- which is breaking test isolation :sweat: -- and adding describe blocks would change the order of execution and break the tests :scream:

lutangar committed:
chore(test): consistent test name

Co-authored-by: Bastien Dumont <32489032+bastiendmt@users.noreply.github.com>

Update apps/backend-e2e/src/e2e/auth.cy.ts

Co-authored-by: Bastien Dumont <32489032+bastiendmt@users.noreply.github.com>

Update apps/backend-e2e/src/e2e/auth.cy.ts

Co-authored-by: Bastien Dumont <32489032+bastiendmt@users.noreply.github.com>

lutangar commented:
@bastiendmt raised interesting concerns about the way our tests are written and I'd like to spend some more time rewriting them. 
We should at least open an issue on the matter!
@aangelot would you consider giving such an issue a priority ?

aangelot commented:
@bastiendmt thank you very much for the reviews

@lutangar so far the goal is to have a usable tool. Secondly, we must respond to issues specific to Industry Map. Then we can focus on testing (and Kepler migration), so that would be November or something. Unless you tell me it's just an hour or it's an emergency.

lutangar committed:
chore(ci): fix running e2e backend tests during CI

chore(ci): test

chore(ci): fix credentials

chore(ci): run background idomitc way

chore(ci): specify container

chore(ci): map to host

chore(ci): setup users

chore(ci): add & use wait-on

chore(ci): run wait on sequentially

chore(ci): run wait on sin server step

chore(env): fix in ci

chore(env): simplify

lutangar committed:
chore(env): fix editor email env variable name
